[AMDGPU] Handle amdgpu.last.use metadata

mbrkusanin opened the pull request:
Convert !amdgpu.last.use metadata into MachineMemOperand for last use
and handle it in SIMemoryLegalizer similar to nontemporal and volatile.

mbrkusanin committed:
[AMDGPU] Handle amdgpu.last.use metadata

Convert !amdgpu.last.use metadata into Machine Memory Operand for last use
and handle it in SIMemoryLegalizer similary to nontemporal and volatile.

mbrkusanin commented:
First commit is part of: https://github.com/llvm/llvm-project/pull/83814
Second and third commits are part of: https://github.com/llvm/llvm-project/pull/83815

mbrkusanin committed:
clang-format

jayfoad commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -2415,6 +2442,16 @@ bool SIGfx12CacheControl::enableVolatileAndOrNonTemporal(
+                                        bool IsLastUse) const {
+  assert(MI.mayLoad() && !MI.mayStore());
+
+  if (IsLastUse && !isScope(MI, AMDGPU::CPol::SCOPE_SYS))
```
What's the reason for not doing this for SCOPE_SYS?

jayfoad replied:
For simplicity I think we should probably set th=3 regardless of scope.

If I understand correctly "last use" does not really mean anything useful for a system scope load. Normally "last use" means you can discard dirty data from any cache at the same or higher level than the scope. But for system scope, there _is_ no cache at that scope, so there is nothing to do. But it is harmless to set th=3 anyway. @Pierre-vh

mbrkusanin committed:
Allow th=3 for SCOPE_SYS

jayfoad reviewed (commented):
Looks good overall. I think we should document `amdgpu.last.use` in AMDGPUUsage, in a new "LLVM IR Metadata" section like the existing "LLVM IR Intrinsics" documentation.

mbrkusanin commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -2471,12 +2491,19 @@ bool SIMemoryLegalizer::expandLoad(const SIMemOpInfo &MOI,
                                                 SIMemOp::LOAD, MOI.isVolatile(),
                                                 MOI.isNonTemporal());
+
+  Changed |= CC->enableLastUse(Inst, MOI.isLastUse());
```
It could still be overridden by nontemporal.

jayfoad replied:
Agreed, lastuse should take priority over nontemporal. (Do you have any tests for that?) I think this would be more obvious if you moved the lastuse logic into enableVolatileAndOrNonTemporal, instead of having a separate enableLastUse, but I won't insist.

mbrkusanin replied:
I moved it to enableVolatileAndOrNonTemporal but had to update methods in all subclasses as well.

arsenm commented on llvm/test/CodeGen/AMDGPU/memory-legalizer-flat-lastuse.ll:
```diff
@@ -0,0 +1,66 @@
+; GFX12-NEXT:    flat_store_b32 v[0:1], v2
+; GFX12-NEXT:    s_endpgm
+entry:
+  %val = load volatile i32, ptr %in, align 4, !amdgpu.last.use !{}
```
need test with interaction with nontemporal?

mbrkusanin replied:
added

jayfoad commented on llvm/test/CodeGen/AMDGPU/memory-legalizer-private-lastuse.ll:
```diff
@@ -0,0 +1,70 @@
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx1200 < %s | FileCheck --check-prefixes=GFX12,GFX12-WGP %s
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx1200 -mattr=+cumode < %s | FileCheck --check-prefixes=GFX12,GFX12-CU %s
+
+define amdgpu_kernel void @private_nontemporal_load_0(ptr addrspace(5) %in, ptr addrspace(1) %out) {
```
Replace "nontemporal" with "last_use" in this file.

mbrkusanin committed:
Docs; new tests + renames; refactor enableLastUse

mbrkusanin commented:
- Rebase
- Updated docs (AMDGPU Usage)
- Fixed test name. Added tests with last.use + nontemporal
- Removed enableLastUse (now it is part of enable enableVolatileAndOrNonTemporal)

mbrkusanin commented on llvm/test/CodeGen/AMDGPU/memory-legalizer-private-lastuse.ll:
```diff
@@ -0,0 +1,70 @@
+  ret void
+}
+
+define amdgpu_kernel void @private_nontemporal_load_1(ptr addrspace(5) %in, ptr addrspace(1) %out) {
```
Forgot to rename it. It's updated now.

jayfoad reviewed (approved):
LGTM, thanks. Please wait until tomorrow for any other comments.
